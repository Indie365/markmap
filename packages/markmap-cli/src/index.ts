import { readFile, writeFile } from 'fs/promises';
import { resolve } from 'path';
import { fileURLToPath } from 'url';
import { Command } from 'commander';
import open from 'open';
import updateNotifier from 'update-notifier';
import { readPackageUp } from 'read-package-up';
import { CSSItem, JSItem, buildJSItem } from 'markmap-common';
import {
  Transformer,
  baseJsPaths,
  type IMarkmapCreateOptions,
  type IAssets,
} from 'markmap-lib';
import {
  ASSETS_PREFIX,
  IDevelopOptions,
  addToolbar,
  assetsDirPromise,
  localProvider,
} from './util';
import { develop } from './dev-server';
import { fetchAssets } from './fetch-assets';
import nodeHtmlToImage from 'node-html-to-image';

export * from 'markmap-lib';
export { develop, fetchAssets };

async function loadFile(path: string) {
  if (path.startsWith(ASSETS_PREFIX)) {
    const relpath = path.slice(ASSETS_PREFIX.length);
    const assetsDir = await assetsDirPromise;
    return readFile(resolve(assetsDir, relpath), 'utf8');
  }
  const res = await fetch(path);
  if (!res.ok) throw res;
  return res.text();
}

async function inlineAssets(assets: IAssets): Promise<IAssets> {
  const [scripts, styles] = await Promise.all([
    Promise.all(
      (assets.scripts || []).map(
        async (item): Promise<JSItem> =>
          item.type === 'script' && item.data.src
            ? {
                type: 'script',
                data: {
                  textContent: await loadFile(item.data.src),
                },
              }
            : item,
      ),
    ),
    Promise.all(
      (assets.styles || []).map(
        async (item): Promise<CSSItem> =>
          item.type === 'stylesheet'
            ? {
                type: 'style',
                data: await loadFile(item.data.href),
              }
            : item,
      ),
    ),
  ]);
  return {
    scripts,
    styles,
  };
}

export async function createMarkmap(
  options: IMarkmapCreateOptions & IDevelopOptions,
): Promise<void> {
  const transformer = new Transformer();
  if (options.offline) {
    transformer.urlBuilder.setProvider('local', localProvider);
    transformer.urlBuilder.provider = 'local';
  } else {
    try {
      await transformer.urlBuilder.findFastestProvider();
    } catch {
      // ignore
    }
  }
  const { root, features, frontmatter } = transformer.transform(
    options.content || '',
  );
  let assets = transformer.getUsedAssets(features);
  assets = {
    ...assets,
    scripts: [
      ...baseJsPaths
        .map((path) => transformer.urlBuilder.getFullUrl(path))
        .map((path) => buildJSItem(path)),
      ...(assets.scripts || []),
    ],
  };
  if (options.toolbar) {
    assets = addToolbar(transformer.urlBuilder, assets);
  }
  if (options.offline) {
    assets = await inlineAssets(assets);
  }
  const html = transformer.fillTemplate(root, assets, {
    baseJs: [],
    jsonOptions: (frontmatter as any)?.markmap,
  });
  const output = options.output || 'markmap.html';
  // here we read html file to image
  nodeHtmlToImage({
    output: options.outputImage,
    html: html,
  });
  await writeFile(output, html, 'utf8');
  if (options.open) open(output);
}

export async function main() {
  const pkg = (
    await readPackageUp({
      cwd: fileURLToPath(import.meta.url),
    })
  )?.packageJson;
  if (!pkg) throw new Error('package.json not found');

  const notifier = updateNotifier({ pkg });
  notifier.notify();

  const program = new Command();
  program
    .version(pkg.version)
    .description('Create a markmap from a Markdown input file')
    .arguments('<input>')
    .option('--no-open', 'do not open the output file after generation')
    .option('--no-toolbar', 'do not show toolbar')
    .option('-o, --output <output>', 'specify filename of the output HTML')
    .option(
      '-O, --output-image <output>',
      'specify filename of the output image',
    )
    .option(
      '--offline',
      'Inline all assets to allow the generated HTML to work offline',
    )
    .option(
      '-w, --watch',
      'watch the input file and update output on the fly, note that this feature is for development only',
    )
    .action(async (input, cmd) => {
      let { offline } = cmd;
      if (cmd.watch) offline = true;
      if (offline) await fetchAssets();
      const content = await readFile(input, 'utf8');
      const output = cmd.output || `${input.replace(/\.\w*$/, '')}.html`;
      const outputImage =
        cmd.outputImage || `${input.replace(/\.\w*$/, '')}.png`;
      if (cmd.watch) {
        await develop(input, {
          open: cmd.open,
          toolbar: cmd.toolbar,
          offline,
        });
      } else {
        await createMarkmap({
          content,
          output,
          outputImage,
          open: cmd.open,
          toolbar: cmd.toolbar,
          offline,
        });
      }
    });
  program.parse(process.argv);
}
