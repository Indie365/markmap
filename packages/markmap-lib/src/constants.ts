export const template = process.env.TEMPLATE || '';

export const baseJsPaths = [
  `d3@${process.env.D3_VERSION}/dist/d3.min.js`,
  `markmap-view@${process.env.VIEW_VERSION}/dist/browser/index.js`,
];

export const svg_marked = `<svg width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"/></svg>`;
export const svg_unmarked = `<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path fill-rule="evenodd" d="M6 5a1 1 0 0 0-1 1v12a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V6a1 1 0 0 0-1-1zM3 6a3 3 0 0 1 3-3h12a3 3 0 0 1 3 3v12a3 3 0 0 1-3 3H6a3 3 0 0 1-3-3v-5z" clip-rule="evenodd"/></svg>`;
