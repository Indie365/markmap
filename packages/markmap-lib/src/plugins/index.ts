import katex from './katex';
import frontmatter from './frontmatter';
import npmUrl from './npm-url';
import hljs from './hljs';
import checkbox from './checkbox';

export * from './base';
export const plugins = [frontmatter, katex, hljs, npmUrl, checkbox];
