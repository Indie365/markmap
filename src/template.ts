const template: string = process.env.TEMPLATE;

export function fillTemplate(data: any): string {
  const html = template
    .replace('{/* data */}', JSON.stringify(data));
  return html;
}
