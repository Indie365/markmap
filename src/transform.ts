import { Remarkable } from 'remarkable';
import { INode, IValue } from './types';

const md = new Remarkable();
md.block.ruler.enable([
  'deflist',
]);

function shallowEqual(a, b): boolean {
  a = a || {};
  b = b || {};
  return Object.keys(a).length === Object.keys(b).length && Object.keys(a).every(k => a[k] === b[k]);
}

function extractInline(token): IValue[] {
  const root: IValue = {
    t: 'inline',
    c: [],
  };
  const stack = [root];
  let style = {};
  for (const child of token.children) {
    const current = stack[stack.length - 1];
    if (child.type === 'text') {
      current.c.push({
        t: 'text',
        v: child.content,
        p: { style },
      });
    } else if (child.type === 'softbreak') {
      current.c.push({
        t: 'softbreak',
      });
    } else if (child.type.endsWith('_open')) {
      const type = child.type.slice(0, -5);
      if (type === 'link') {
        const item = {
          t: 'link',
          c: [],
          p: {
            href: child.href,
            title: child.title,
          },
        };
        current.c.push(item);
        stack.push(item);
      } else {
        style = {
          ...style,
          [type]: true,
        };
      }
    } else if (child.type.endsWith('_close')) {
      const type = child.type.slice(0, -6);
      if (type === 'link') {
        stack.pop();
      } else {
        style = {
          ...style,
          [type]: false,
        };
      }
    }
  }
  return root.c;
}

function cleanNode(node: INode, depth = 0): void {
  if (node.t === 'heading') {
    // drop all paragraphs
    node.c = node.c.filter(item => item.t !== 'paragraph');
  } else if (node.t === 'list_item') {
    // keep first paragraph as content of list_item, drop others
    node.c = node.c.filter(item => {
      if (item.t === 'paragraph') {
        if (!node.v.length) node.v.push(...item.v);
        return false;
      }
      return true;
    });
    if (node.p?.index != null) {
      node.v.unshift({
        t: 'text',
        v: `${node.p.index}. `,
      });
    }
  } else if (node.t === 'ordered_list') {
    let index = node.p?.start ?? 1;
    node.c.forEach(item => {
      if (item.t === 'list_item') {
        item.p = {
          ...item.p,
          index: index,
        };
        index += 1;
      }
    });
  }
  if (node.c.length === 0) {
    delete node.c;
  } else {
    if (node.c.length === 1 && !node.c[0].v.length) {
      node.c = node.c[0].c;
    }
    node.c.forEach(child => cleanNode(child, depth + 1));
  }
  let last: IValue;
  const content = [];
  for (const item of node.v) {
    if (last?.t === 'text' && item.t === 'text' && shallowEqual(last.p?.style, item.p?.style)) {
      last.v += item.v;
    } else {
      content.push(item);
    }
    last = item;
  }
  node.v = content;
  node.d = depth;
  delete node.p;
}

export function buildTree(tokens): INode {
  // TODO deal with <dl><dt>
  const root: INode = {
    t: 'root',
    d: 0,
    v: [],
    c: [],
  };
  const stack = [root];
  let depth = 0;
  for (const token of tokens) {
    let current = stack[stack.length - 1];
    if (token.type.endsWith('_open')) {
      const type = token.type.slice(0, -5);
      const payload: any = {};
      if (type === 'heading') {
        depth = token.hLevel;
        while (current?.d >= depth) {
          stack.pop();
          current = stack[stack.length - 1];
        }
      } else {
        depth = Math.max(depth, current?.d || 0) + 1;
        if (type === 'ordered_list') {
          payload.start = token.order;
        }
      }
      const item: INode = {
        t: type,
        d: depth,
        v: [],
        p: payload,
        c: [],
      };
      current.c.push(item);
      stack.push(item);
    } else if (!current) {
      continue
    } else if (token.type === `${current.t}_close`) {
      if (current.t === 'heading') {
        depth = current.d;
      } else {
        stack.pop();
        depth = 0;
      }
    } else if (token.type === 'inline') {
      current.v.push(...extractInline(token));
    } else {
      // ignore other nodes
    }
  }
  return root;
}

export function transform(content: string): INode {
  const tokens = md.parse(content || '', {});
  let root = buildTree(tokens);
  cleanNode(root);
  if (root.c?.length === 1) root = root.c[0];
  return root;
}
