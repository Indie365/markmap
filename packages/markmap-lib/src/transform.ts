import {
  CSSItem,
  IMarkmapOptions,
  IPureNode,
  JSItem,
  UrlBuilder,
  buildJSItem,
  persistCSS,
  persistJS,
  wrapFunction,
} from 'markmap-common';
import { IHtmlParserOptions, buildTree } from 'markmap-html-parser';
import { Remarkable } from 'remarkable';
import { baseJsPaths, template } from './constants';
import { plugins as builtInPlugins, createTransformHooks } from './plugins';
import {
  IAssets,
  IFeatures,
  ITransformContext,
  ITransformHooks,
  ITransformPlugin,
  ITransformResult,
  ITransformer,
  IMarkmapJSONOptions,
} from './types';
import { patchCSSItem, patchJSItem } from './util';

export { builtInPlugins };

function cleanNode(node: IPureNode): IPureNode {
  while (!node.content && node.children.length === 1) {
    node = node.children[0];
  }
  while (node.children.length === 1 && !node.children[0].content) {
    node = {
      ...node,
      children: node.children[0].children,
    };
  }
  return {
    ...node,
    children: node.children.map(cleanNode),
  };
}

export class Transformer implements ITransformer {
  hooks: ITransformHooks;

  md: Remarkable;

  assetsMap: Record<string, IAssets> = {};

  urlBuilder = new UrlBuilder();

  plugins: ITransformPlugin[];

  constructor(
    plugins: Array<
      ITransformPlugin | (() => ITransformPlugin)
    > = builtInPlugins,
  ) {
    this.hooks = createTransformHooks(this);
    this.plugins = plugins.map((plugin) =>
      typeof plugin === 'function' ? plugin() : plugin,
    );

    const assetsMap: typeof this.assetsMap = {};
    for (const { name, transform } of this.plugins) {
      assetsMap[name] = transform(this.hooks);
    }
    this.assetsMap = assetsMap;

    const md = new Remarkable('full', {
      html: true,
      breaks: true,
      maxNesting: Infinity,
    } as Remarkable.Options);
    md.renderer.rules.htmltag = wrapFunction(
      md.renderer.rules.htmltag,
      (render, ...args) => {
        const result = render(...args);
        this.hooks.htmltag.call({ args, result });
        return result;
      },
    );
    this.md = md;
    this.hooks.parser.call(md);
  }

  transform(
    content: string,
    opts?: Partial<IHtmlParserOptions>,
  ): ITransformResult {
    const context: ITransformContext = {
      content,
      features: {},
      contentLineOffset: 0,
    };
    this.hooks.beforeParse.call(this.md, context);
    const html = this.md.render(context.content, {});
    this.hooks.afterParse.call(this.md, context);
    const root = cleanNode(
      buildTree(html, {
        ...context.frontmatter?.markmap?.htmlParser,
        ...opts,
      }),
    );
    return { ...context, root };
  }

  /**
   * Get all assets from enabled plugins or filter them by plugin names as keys.
   */
  getAssets(keys?: string[]): IAssets {
    const styles: CSSItem[] = [];
    const scripts: JSItem[] = [];
    keys ??= this.plugins.map((plugin) => plugin.name);
    for (const assets of keys.map((key) => this.assetsMap[key])) {
      if (assets) {
        if (assets.styles) styles.push(...assets.styles);
        if (assets.scripts) scripts.push(...assets.scripts);
      }
    }
    return {
      styles: styles.map((item) => patchCSSItem(this.urlBuilder, item)),
      scripts: scripts.map((item) => patchJSItem(this.urlBuilder, item)),
    };
  }

  /**
   * Get used assets by features object returned by `transform`.
   */
  getUsedAssets(features: IFeatures): IAssets {
    const keys = this.plugins
      .map((plugin) => plugin.name)
      .filter((name) => features[name]);
    return this.getAssets(keys);
  }

  fillTemplate(
    root: IPureNode | null,
    assets: IAssets,
    extra?: {
      baseJs?: JSItem[];
      jsonOptions?: IMarkmapJSONOptions;
      getOptions?: (
        jsonOptions: IMarkmapJSONOptions,
      ) => Partial<IMarkmapOptions>;
    },
  ): string {
    extra = {
      ...extra,
    };
    extra.baseJs ??= baseJsPaths
      .map((path) => this.urlBuilder.getFullUrl(path))
      .map((path) => buildJSItem(path));
    const { scripts, styles } = assets;
    const cssList = [...(styles ? persistCSS(styles) : [])];
    const context = {
      getMarkmap: () => window.markmap,
      getOptions: extra.getOptions,
      jsonOptions: extra.jsonOptions,
      root,
    };
    const jsList = [
      ...persistJS(
        [
          ...extra.baseJs,
          ...(scripts || []),
          {
            type: 'iife',
            data: {
              fn: (
                getMarkmap: (typeof context)['getMarkmap'],
                getOptions: (typeof context)['getOptions'],
                root: (typeof context)['root'],
                jsonOptions: IMarkmapJSONOptions,
              ) => {
                const markmap = getMarkmap();
                window.mm = markmap.Markmap.create(
                  'svg#mindmap',
                  (getOptions || markmap.deriveOptions)(jsonOptions),
                  root,
                );
              },
              getParams: ({ getMarkmap, getOptions, root, jsonOptions }) => {
                return [getMarkmap, getOptions, root, jsonOptions];
              },
            },
          } as JSItem,
        ],
        context,
      ),
    ];
    const html = template
      .replace('<!--CSS-->', () => cssList.join(''))
      .replace('<!--JS-->', () => jsList.join(''));
    return html;
  }
}
