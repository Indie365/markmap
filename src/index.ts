import { promises as fs } from 'fs';
import open from 'open';
import { transform } from './transform';
import { fillTemplate } from './template';
import { IMarkmapCreateOptions } from './types';

export async function createMarkmap(options: IMarkmapCreateOptions = {}): Promise<void> {
  const {
    input,
    open: openFile = true,
    ...rest
  } = options;
  let { content, output } = options;
  if (input) {
    content = await fs.readFile(input, 'utf8');
  }
  if (!output) {
    output = input ? `${input.replace(/\.\w*$/, '')}.html` : 'markmap.html';
  }

  const root = transform(content || '');
  const html = fillTemplate(root, rest);
  fs.writeFile(output, html, 'utf8');
  if (openFile) open(output);
}
