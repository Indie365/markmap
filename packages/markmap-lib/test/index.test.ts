import { Transformer } from '../src/index';

test('plugins', () => {
  const transformer = new Transformer();
  expect(transformer.plugins.map((plugin) => plugin.name)).toEqual([
    'frontmatter',
    'katex',
    'hljs',
    'npmUrl',
    'checkbox',
  ]);
  const assets = transformer.getAssets();
  expect(assets).toMatchSnapshot();
});

test('custom url provider', () => {
  const transformer = new Transformer();
  transformer.urlBuilder.setProvider('local', (path) => `/local/${path}`);
  transformer.urlBuilder.provider = 'local';
  const assets = transformer.getAssets();
  expect(assets).toMatchSnapshot();
});

test('content without frontmatter', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
- l1
  - l1.1
  - l1.2
    - l1.2.1
`);
  expect(result).toMatchSnapshot();
});

test('content with frontmatter', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
---
markmap:
  color: blue
---

- l1
  - l1.1
  - l1.2
    - l1.2.1
`);
  expect(result).toMatchSnapshot();
});

test('content with line endings of CRLF', () => {
  const transformer = new Transformer();
  const result = transformer.transform(
    `\
---
markmap:
  color: blue
---

- l1
  - l1.1
  - l1.2
    - l1.2.1
`.replace(/\n/g, '\r\n'),
  );
  expect(result).toMatchSnapshot();
});

test('content with only katex enabled', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
---
markmap:
  color: blue
---

- $x = {-b \pm \sqrt{b^2-4ac} \over 2a}$
`);
  expect(result).toMatchSnapshot();
  expect(transformer.getUsedAssets(result.features)).toMatchSnapshot();
});

test('tables', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
| products | price |
|-|-|
| apple | 10 |
| banana | 12 |
`);
  expect(result).toMatchSnapshot();
});

test('images', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
![](image1.png)

![](image2.png)
`);
  expect(result).toMatchSnapshot();
});

test('checkboxes', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
# Housework

## Main

- [x] Dishes
- [ ] Cleaning the bathroom
- [x] Change the light bulbs
- [ ] something else

## [x] should it works on titles?

## [x] idk if it should!

### [ ] test

### [x] test

- [x] test
- [x] test


## [x] only works on list items is better
\`\`\`
[ ] this is not a checkbox either
\`\`\`
`);
  expect(result).toMatchSnapshot();
});