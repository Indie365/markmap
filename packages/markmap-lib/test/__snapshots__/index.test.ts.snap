// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkboxes 1`] = `
{
  "content": "# Housework

## Main

- [x] Dishes
- [ ] Cleaning the bathroom
- [x] Change the light bulbs
- [ ] something else

## [x] should it works on titles?

## [x] idk if it should!

### [ ] test

### [x] test

- [x] test
- [x] test


## [x] only works on list items is better
\`\`\`
[ ] this is not a checkbox either
\`\`\`
  ",
  "contentLineOffset": 0,
  "features": {
    "hljs": true,
  },
  "root": {
    "children": [
      {
        "children": [
          {
            "children": [],
            "content": "<svg width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg>  Dishes",
          },
          {
            "children": [],
            "content": "<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path fill-rule="evenodd" d="M6 5a1 1 0 0 0-1 1v12a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V6a1 1 0 0 0-1-1zM3 6a3 3 0 0 1 3-3h12a3 3 0 0 1 3 3v12a3 3 0 0 1-3 3H6a3 3 0 0 1-3-3v-5z" clip-rule="evenodd"></path></svg>  Cleaning the bathroom",
          },
          {
            "children": [],
            "content": "<svg width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg>  Change the light bulbs",
          },
          {
            "children": [],
            "content": "<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path fill-rule="evenodd" d="M6 5a1 1 0 0 0-1 1v12a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V6a1 1 0 0 0-1-1zM3 6a3 3 0 0 1 3-3h12a3 3 0 0 1 3 3v12a3 3 0 0 1-3 3H6a3 3 0 0 1-3-3v-5z" clip-rule="evenodd"></path></svg>  something else",
          },
        ],
        "content": "Main",
      },
      {
        "children": [],
        "content": "<svg width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg>  should it works on titles?",
      },
      {
        "children": [
          {
            "children": [],
            "content": "<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path fill-rule="evenodd" d="M6 5a1 1 0 0 0-1 1v12a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V6a1 1 0 0 0-1-1zM3 6a3 3 0 0 1 3-3h12a3 3 0 0 1 3 3v12a3 3 0 0 1-3 3H6a3 3 0 0 1-3-3v-5z" clip-rule="evenodd"></path></svg>  test",
          },
          {
            "children": [
              {
                "children": [],
                "content": "<svg width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg>  test",
              },
              {
                "children": [],
                "content": "<svg width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg>  test",
              },
            ],
            "content": "<svg width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg>  test",
          },
        ],
        "content": "<svg width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg>  idk if it should!",
      },
      {
        "children": [
          {
            "children": [],
            "content": "<pre><code><span class="hljs-string">[ ]</span> <span class="hljs-keyword">this</span> <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> a checkbox either
</code></pre>",
          },
        ],
        "content": "<svg width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg>  only works on list items is better",
      },
    ],
    "content": "Housework",
  },
}
`;

exports[`content with frontmatter 1`] = `
{
  "content": "
- l1
  - l1.1
  - l1.2
    - l1.2.1
",
  "contentLineOffset": 4,
  "features": {},
  "frontmatter": {
    "markmap": {
      "color": [
        "blue",
      ],
    },
  },
  "root": {
    "children": [
      {
        "children": [],
        "content": "l1.1",
      },
      {
        "children": [
          {
            "children": [],
            "content": "l1.2.1",
          },
        ],
        "content": "l1.2",
      },
    ],
    "content": "l1",
  },
}
`;

exports[`content with line endings of CRLF 1`] = `
{
  "content": "
- l1
  - l1.1
  - l1.2
    - l1.2.1
",
  "contentLineOffset": 4,
  "features": {},
  "frontmatter": {
    "markmap": {
      "color": [
        "blue",
      ],
    },
  },
  "root": {
    "children": [
      {
        "children": [],
        "content": "l1.1",
      },
      {
        "children": [
          {
            "children": [],
            "content": "l1.2.1",
          },
        ],
        "content": "l1.2",
      },
    ],
    "content": "l1",
  },
}
`;

exports[`content with only katex enabled 1`] = `
{
  "content": "
- $x = {-b pm sqrt{b^2-4ac} over 2a}$
",
  "contentLineOffset": 4,
  "features": {
    "katex": true,
  },
  "frontmatter": {
    "markmap": {
      "color": [
        "blue",
      ],
    },
  },
  "root": {
    "children": [],
    "content": "<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>x</mi><mo>=</mo><mrow><mo>−</mo><mi>b</mi><mi>p</mi><mi>m</mi><mi>s</mi><mi>q</mi><mi>r</mi><mi>t</mi><mrow><msup><mi>b</mi><mn>2</mn></msup><mo>−</mo><mn>4</mn><mi>a</mi><mi>c</mi></mrow><mi>o</mi><mi>v</mi><mi>e</mi><mi>r</mi><mn>2</mn><mi>a</mi></mrow></mrow><annotation encoding="application/x-tex">x = {-b pm sqrt{b^2-4ac} over 2a}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.4306em;"></span><span class="mord mathnormal">x</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:1.0085em;vertical-align:-0.1944em;"></span><span class="mord"><span class="mord">−</span><span class="mord mathnormal">b</span><span class="mord mathnormal">p</span><span class="mord mathnormal">m</span><span class="mord mathnormal">s</span><span class="mord mathnormal" style="margin-right:0.03588em;">q</span><span class="mord mathnormal" style="margin-right:0.02778em;">r</span><span class="mord mathnormal">t</span><span class="mord"><span class="mord"><span class="mord mathnormal">b</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mord">4</span><span class="mord mathnormal">a</span><span class="mord mathnormal">c</span></span><span class="mord mathnormal">o</span><span class="mord mathnormal" style="margin-right:0.03588em;">v</span><span class="mord mathnormal" style="margin-right:0.02778em;">er</span><span class="mord">2</span><span class="mord mathnormal">a</span></span></span></span></span>",
  },
}
`;

exports[`content with only katex enabled 2`] = `
{
  "scripts": [
    {
      "data": {
        "fn": [Function],
        "getParams": [Function],
      },
      "type": "iife",
    },
    {
      "data": {
        "defer": true,
        "src": "https://cdn.jsdelivr.net/npm/webfontloader@undefined/webfontloader.js",
      },
      "type": "script",
    },
  ],
  "styles": [
    {
      "data": {
        "href": "https://cdn.jsdelivr.net/npm/katex@undefined/dist/katex.min.css",
      },
      "type": "stylesheet",
    },
  ],
}
`;

exports[`content without frontmatter 1`] = `
{
  "content": "- l1
  - l1.1
  - l1.2
    - l1.2.1
",
  "contentLineOffset": 0,
  "features": {},
  "root": {
    "children": [
      {
        "children": [],
        "content": "l1.1",
      },
      {
        "children": [
          {
            "children": [],
            "content": "l1.2.1",
          },
        ],
        "content": "l1.2",
      },
    ],
    "content": "l1",
  },
}
`;

exports[`custom url provider 1`] = `
{
  "scripts": [
    {
      "data": {
        "fn": [Function],
        "getParams": [Function],
      },
      "type": "iife",
    },
    {
      "data": {
        "defer": true,
        "src": "/local/webfontloader@undefined/webfontloader.js",
      },
      "type": "script",
    },
  ],
  "styles": [
    {
      "data": {
        "href": "/local/katex@undefined/dist/katex.min.css",
      },
      "type": "stylesheet",
    },
    {
      "data": {
        "href": "/local/@highlightjs/cdn-assets@undefined/styles/default.min.css",
      },
      "type": "stylesheet",
    },
  ],
}
`;

exports[`images 1`] = `
{
  "content": "![](image1.png)

![](image2.png)
",
  "contentLineOffset": 0,
  "features": {},
  "root": {
    "children": [
      {
        "children": [],
        "content": "<img src="image1.png" alt="">",
      },
      {
        "children": [],
        "content": "<img src="image2.png" alt="">",
      },
    ],
    "content": "",
  },
}
`;

exports[`plugins 1`] = `
{
  "scripts": [
    {
      "data": {
        "fn": [Function],
        "getParams": [Function],
      },
      "type": "iife",
    },
    {
      "data": {
        "defer": true,
        "src": "https://cdn.jsdelivr.net/npm/webfontloader@undefined/webfontloader.js",
      },
      "type": "script",
    },
  ],
  "styles": [
    {
      "data": {
        "href": "https://cdn.jsdelivr.net/npm/katex@undefined/dist/katex.min.css",
      },
      "type": "stylesheet",
    },
    {
      "data": {
        "href": "https://cdn.jsdelivr.net/npm/@highlightjs/cdn-assets@undefined/styles/default.min.css",
      },
      "type": "stylesheet",
    },
  ],
}
`;

exports[`tables 1`] = `
{
  "content": "| products | price |
|-|-|
| apple | 10 |
| banana | 12 |
",
  "contentLineOffset": 0,
  "features": {},
  "root": {
    "children": [],
    "content": "<table>
<thead>
<tr><th>products</th><th>price</th></tr>
</thead>
<tbody>
<tr><td>apple</td><td>10</td></tr>
<tr><td>banana</td><td>12</td></tr>
</tbody>
</table>",
  },
}
`;
